### ISSUE TITLE
Reordering of *ngFor list corrupts the current state of simple animations contained inside the item template.

### ISSUE THREAD

**Issue by simeyla** [angular/angular#50861]
### Which @angular/* package(s) are the source of the bug?

animations

### Is this a regression?

No

### Description

Setup:

- I have a list of products in a shopping cart, each with an Add button. 
- Products are displayed in an *ngFor list (with a trackBy function)
- When an item is 'Added',  the 'Add' button fades out using a simple opacity animation. (not with an *ngIf)
- At the same time a 'Remove' button is faded in.
- When the item is 'Removed' the buttons return to their original state.

Problem: 
- When the parent list of products is reordered (click 'Sort by Name') the simple opacity animation for the buttons gets broken (the current visible or hidden state appears to get corrupted or reset to void).
- By 'broken' I mean that the transition is not triggered from `visible => hidden` and the opacity changes instantly.

Important: This is NOT an `:enter` / `:leave` animation issue like https://github.com/angular/angular/issues/40752

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/issue-50861-v2?file=src%2Fmain.ts

Note: This link has been updated to a version using a new component for each product in the list.

### Please provide the exception or error you saw

```true
The animation doesn't work immediately after the list is reordered.
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 16.0.6
Node: 18.14.0
Package Manager: npm 9.3.1
OS: win32 x64

Angular: 16.0.0
... animations, common, compiler, core, forms, platform-browser
... platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1601.1
@angular-devkit/build-angular   16.1.1
@angular-devkit/core            16.1.1
@angular-devkit/schematics      16.0.6
@angular/cli                    16.0.6
@angular/compiler-cli           16.1.2
@schematics/angular             16.0.6
rxjs                            7.8.1
typescript                      5.0.4
```

### Anything else?

I want to propose a hypothesis for what seems to be happening:

The Add / Remove buttons being animated are simple Angular opacity animations (not hidden with *ngIf). 

    <button [@visibilityAnimation]="(product.qty == 0) ? 'visible' : 'hidden'" 
            (click)="addToCart(product, 1)">Add to cart</button>

The animation is very simple:

```
export const visibilityAnimation = trigger('visibilityAnimation', 
[
  state('hidden', style({ opacity: 0 })),
  state('visible', style({ opacity: 1 })),

  // fade in
  transition('hidden => visible', [animate('1s')]),

  // fade out 
  transition('visible => hidden', [animate('.5s')]),
]);
```

So clearly the animation state can only ever be `'visible'` or `'hidden`' and one of the defined transitions should always work.

Indeed, before the list has been sorted, when adding an item, the `visible => hidden` transition is triggered as expected every time. 

After sorting the list - the `visible => hidden` or `hidden => visible` transitions aren't triggered on the next click (the button doesn't fade). However it isn't completely broken - once you click again it will work. 

It is as if the state of each animation is reset to a 'null' or 'void' state. In fact if you change the transitions to be `* => hidden` and `* => visible` then the animation always works. (This is unacceptable in reality since I don't want the animation to run when the page is loaded).

It may be possible to fix this by making the list items separate components, but that doesn't work well for my actual situation.

As mentioned above, any similar existing issues I was able to find are related to the *ngIf (:enter / :leave) animation mechanism which does not apply here. Whether or not they are related under the hood I have no way of knowing.

**Comment by simeyla** [comment]
I've discovered some additional information during testing today.

- My original repro was a single file which demonstrated the problem but did not prove the wrong state was set on a product.
- I have now converted the repro to a more realistic structure that matches my actual application:

    `AppComponent` > `ProductListComponent` > List of `ProductComponent` instances in an `*ngFor`

- Load this version and immediately click `Sort by Name`.
https://stackblitz.com/edit/issue-50861-v2?file=src%2Fmain.ts

- You will see pink outlines around the buttons. This is a the animation state `void` (defined in `animations.ts). This demonstrates that the animation state for `@visibilityAnimation` is broken. I was unable to duplicate this with a single file - it required a real component tree.

- Now load this version and immediately click `Sort by Name`
https://stackblitz.com/edit/issue-50861-v3-mslhcq?file=src%2Fproduct-list%2Fproduct-list.component.html

The pink outline isn't there, but the animation transition is broken if you subsequently click on 'Add to Cart' (it switches instantly).

The ONLY difference in the second version is the addition of the following line to `product-list.component`:

   animations: [ visibilityAnimation ]

What's really interesting is that `product-list.component` doesn't even use this animation. It is applied by the children `product.component` instances.

It is as if Angular is looking in the wrong place for metadata about the animation after the list items has been reordered.  Adding the animation definition to `product-list.component` (which owns the `*ngFor`) allows it to be found at least as far as not setting a `void` state is concerned.
